services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java

```diff
@@ -0,0 +1,385 @@
+      .newBuilder()
+      .setStreamsStarted(1)
+      .setLastLocalStreamCreatedTimestamp(Timestamps.fromNanos(2))
+      .setLastRemoteStreamCreatedTimestamp(Timestamps.fromNanos(3))
+      .setStreamsSucceeded(4)
+      .setStreamsFailed(5)
+      .setMessagesSent(6)
+      .setMessagesReceived(7)
+      .setKeepAlivesSent(8)
+      .setLastMessageSentTimestamp(Timestamps.fromNanos(9))
+      .setLastMessageReceivedTimestamp(Timestamps.fromNanos(10))
+      .setLocalFlowControlWindow(Int64Value.newBuilder().setValue(11).build())
+      .setRemoteFlowControlWindow(Int64Value.newBuilder().setValue(12).build())
+      .build();
+  private final Address localAddress = Address
+      .newBuilder()
+      .setTcpipAddress(
+          TcpIpAddress
+              .newBuilder()
+              .setIpAddress(ByteString.copyFrom(
+                  ((InetSocketAddress) socket.local).getAddress().getAddress()))
+              .build())
+      .build();
+  private final Address remoteAddress = Address
+      .newBuilder()
+      .setTcpipAddress(
+          TcpIpAddress
+              .newBuilder()
+              .setIpAddress(ByteString.copyFrom(
+                  ((InetSocketAddress) socket.remote).getAddress().getAddress()))
+              .build())
+      .build();
+
+  @Test
+  public void toChannelRef() {
+    assertEquals(channelRef, ChannelzProtoUtil.toChannelRef(channel));
+  }
+
+  @Test
+  public void toSubchannelRef() {
+    assertEquals(subchannelRef, ChannelzProtoUtil.toSubchannelRef(subchannel));
+  }
+
+  @Test
+  public void toServerRef() {
+    assertEquals(serverRef, ChannelzProtoUtil.toServerRef(server));
+  }
+
+  @Test
+  public void toSocketRef() {
+    assertEquals(socketRef, ChannelzProtoUtil.toSocketRef(socket));
+  }
+
+  @Test
+  public void toState() {
+    assertEquals(State.CONNECTING, ChannelzProtoUtil.toState(ConnectivityState.CONNECTING));
+    assertEquals(State.READY, ChannelzProtoUtil.toState(ConnectivityState.READY));
+    assertEquals(
+        State.TRANSIENT_FAILURE, ChannelzProtoUtil.toState(ConnectivityState.TRANSIENT_FAILURE));
+    assertEquals(State.IDLE, ChannelzProtoUtil.toState(ConnectivityState.IDLE));
+    assertEquals(State.SHUTDOWN, ChannelzProtoUtil.toState(ConnectivityState.SHUTDOWN));
+    assertEquals(State.UNKNOWN, ChannelzProtoUtil.toState(null));
+  }
+
+  @Test
+  public void toSocket() throws Exception {
+    assertEquals(
+        Socket
+            .newBuilder()
+            .setRef(socketRef)
+            .setLocal(localAddress)
+            .setRemote(remoteAddress)
+            .setData(socketData)
+            .build(),
+        ChannelzProtoUtil.toSocket(socket));
+  }
+
+  @Test
+  public void toSocketData() {
+    assertEquals(
+        socketData,
+        ChannelzProtoUtil.toSocketData(socket.transportStats));
+  }
+
+  @Test
+  public void toAddress_inet() throws Exception {
+    InetSocketAddress inet4 = new InetSocketAddress(Inet4Address.getByName("10.0.0.1"), 1000);
```

Comment:
does this work for inprocess?

Reply:
`InProcessTransport` does not keep stats right now, but it will eventually. Specifically for sockets, `ChannelzProtoUtil.toAddress` will turn non UDS, non inet sockets into unknown sockets. So we just need to make sure `InProcessSocketAddress` has a properly working `toString`.